internal/store/firestore.go

```
@@ -0,0 +1,154 @@
+		return nil, fmt.Errorf("unable to fetch hostname document for %s: %s", hostname, err)
+	}
+
+	var h *types.Hostname
+	hostnameDoc.DataTo(&h)
+	return h, nil
+}
+
+func (f *Firestore) ReadHostnameMetadata(ctx context.Context, siteId string, env string, hostname string) (*types.HostnameMetadata, error) {
+	logger := f.logger.WithField("method", "ReadHostnameMetadata")
+
+	collection, err := f.getCollection(ctx, hostnameMetadataCollectionName, logger)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get collection %s: %s", hostnameMetadataCollectionName, err)
+	}
+	hostnameDocument, err := f.getHostnameDocumentBySiteEnv(
+		ctx,
+		collection,
+		siteId,
+		env,
+		hostname)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get hostname document for %s: %s", hostname, err)
+	}
+	var h *types.HostnameMetadata
+	hostnameDocument.DataTo(&h)
+	return h, nil
+}
+
+func (f *Firestore) ReadEdgeLogic(ctx context.Context, siteId string, env string, hostname string) (*types.EdgeLogic, error) {
+	logger := f.logger.WithField("method", "ReadEdgeLogic")
+
+	collection, err := f.getCollection(ctx, edgeLogicCollectionName, logger)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get collection %s: %s", edgeLogicCollectionName, err)
+	}
+
+	hostnameDocument, err := f.getHostnameDocumentBySiteEnv(
+		ctx,
+		collection,
+		siteId,
+		env,
+		hostname,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get hostname focument for %s: %s", hostname, err)
+	}
+
+	var h *types.EdgeLogic
+	hostnameDocument.DataTo(&h)
+	return h, nil
+}
+
+func (*Firestore) getHostnameDocumentBySiteEnv(
+	ctx context.Context,
+	rootCollection *firestore.CollectionRef,
+	siteId string,
+	env string,
+	hostname string) (*firestore.DocumentSnapshot, error) {
+	siteDocRef := rootCollection.Doc(siteId)
+	siteDoc, err := siteDocRef.Get(ctx)
+
+	if err != nil {
+		if status.Code(err) == codes.NotFound {
+			return nil, fmt.Errorf("document not found for site %s %w", siteId, err)
```

For the cloud function maybe it will be better to log the error and return `nil` to avoid additional retry.

---

Since this is the store level, I think returning the error here makes sense, but in the function that calls this, I think you're right. We'll want to consider how to handle this to avoid repeatedly retrying.

---

I mentioned retry policy in a [slack thread](https://pantheon.slack.com/archives/C02LA4DFQUV/p1679668553609099) - as it is written right now, the only spot (unless the code panics, of course) where we let the function die instead of logging errors and returning successfully is if the firestore client cannot be created.